### PR TITLE
Feature 15 ceres

### PR DESCRIPTION
**Pull request by chutsu**
This PR is in relation to #15.

In summary this PR:

- Removes `wave_estimation` (not needed yet) and `wave_kinematics` (Arun is in charge of that)
- Update install script to install opencv 2.4 not 3.0
- Improves install script to hide install outputs to a log file, and only echo the log when an error has occurred
- Add ceres auto, numerical and analytical diff code examples
- Add docs for how to write the three different kinds of cost functions as well as how to load the data
- Add tests to only show that examples can execute, the tests themselves do not do any actual meaningful tests.

### CODE REVIEW

**Comment by chutsu** [comment]
This PR is in relation to #15

**Comment by stevenwaslander** [comment]
This PR is like 9 ideas in one.  Best to decompose so we're not mixing directory reorganizing with substantial commits like Ceres.

**Comment by chutsu** [comment]
Indeed apologies for that, I kept having ideas of how the install scripts would work and OpenCV 3.0 was giving me headaches.

**Comment by stevenwaslander** [comment]
Should be managable, and it's hard to review things that don't work, so initial commits of new modules tend to be large and hard to review. "It gets better."

File: docs/ref/optimizers/ceres_tutorial.md [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+        residual[0] = T(y_) - exp(m[0] * T(x_) + c[0]);
+        return true;
+      }
+
+    private:
+      // Observations for a sample.
+      const double x_;
+      const double y_;
+    };
+
```
Comment (stevenwaslander) [inline docs/ref/optimizers/ceres_tutorial.md]: Are we using this->x or x_, y_ for members? I know this is just docs, so might not need to match.

File: wave_optimization/CMakeLists.txt [hunk 2]
```diff
@@ -5,38 +5,50 @@ IF (CMAKE_COMPILER_IS_GNUCXX)
   ADD_DEFINITIONS(-std=gnu++0x)
 ENDIF()
 
+# PACKAGES
+FIND_PACKAGE(Ceres REQUIRED)
+
 # INCLUDES
```
Comment (stevenwaslander) [inline wave_optimization/CMakeLists.txt]: Is there a way to not depend on Ceres unless you select it to be used? We don't want 100 dependencies to install, in general.  I noticed this with MCPTAM too, still depends on g2o even though Ceres was being used (or vice versa).

File: wave_optimization/scripts/plot_3d_data.py [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+        data["x"].append(float(line[0]))
+        data["y"].append(float(line[1]))
+        data["z"].append(float(line[2]))
+
+    return data
+
+
+def rotation_matrix(q):
+    R_00 = 1.0 - 2.0 * pow(q[1], 2) - 2.0 * pow(q[2], 2)
+    R_01 = 2.0 * q[0] * q[1] + 2.0 * q[3] * q[2]
```
Comment (stevenwaslander) [inline wave_optimization/scripts/plot_3d_data.py]: Why is a rotation matrix defined in a plotting function.  Shouldn't this be standardized?

File: wave_optimization/src/ceres/ceres_examples.cpp [hunk 4]
```diff
@@ -0,0 +1,86 @@
+#include "wave/optimization/ceres/ceres_examples.hpp"
+
+namespace wave {
+namespace ceres {
```
Comment (stevenwaslander) [inline wave_optimization/src/ceres/ceres_examples.cpp]: Seems like a bad idea to have our own namespace called ceres, and then to frequently use the root ceres namespace. Yikes on readability

File: wave_utils/CMakeLists.txt [hunk 5]
```diff
@@ -2,60 +2,59 @@ CMAKE_MINIMUM_REQUIRED(VERSION 2.8.3)
...
+    # TARGET
+    wave_utils_tests
+    # TESTS
+    tests/utils/config_test.cpp
+    tests/utils/data_test.cpp
+    tests/utils/filesystem_test.cpp
+    tests/utils/math_test.cpp
+    tests/utils/time_test.cpp
+    # TEST RUNNER
+    tests/test_runner.cpp
```
Comment (stevenwaslander) [inline wave_utils/CMakeLists.txt]: Are these formatting changes intentional?

File: wave_utils/include/wave/utils/logging.hpp [hunk 6]
```diff
@@ -6,12 +6,9 @@ namespace wave {
...
 
-#define LOG_ERROR(M, ...)              \
-    fprintf(stderr,                    \
-            "[ERROR] [%s:%d] " M "\n", \
-            FILENAME,              \
-            __LINE__,                  \
-            ##__VA_ARGS__)
+#define LOG_ERROR(M, ...) \
+    fprintf(              \
+      stderr, "[ERROR] [%s:%d] " M "\n", FILENAME, __LINE__, ##__VA_ARGS__)
```
Comment (stevenwaslander) [inline wave_utils/include/wave/utils/logging.hpp]: Looks like an unrelated revision, for next time.

Comment (chutsu) [inline wave_utils/include/wave/utils/logging.hpp]: that was automatic from clang-format, some reason it didn't change it till now?

File: wave_utils/CMakeLists.txt (hunk 5 above)
Comment (chutsu) [inline wave_utils/CMakeLists.txt]: Yes, I was trying to match our current 4 space indent rule with the code.

File: wave_optimization/scripts/plot_3d_data.py (hunk 3 above)
Comment (chutsu) [inline wave_optimization/scripts/plot_3d_data.py]: Yikes these scripts are not meant to be here

File: wave_optimization/CMakeLists.txt (hunk 2 above)
Comment (chutsu) [inline wave_optimization/CMakeLists.txt]: We can do that.

File: docs/ref/optimizers/ceres_tutorial.md (hunk 1 above)
Comment (chutsu) [inline docs/ref/optimizers/ceres_tutorial.md]: Updated to match our coding standards

File: wave_optimization/src/ceres/ceres_examples.cpp (hunk 4 above)
Comment (chutsu) [inline wave_optimization/src/ceres/ceres_examples.cpp]: Done, removed ceres and ceres.examples namespace.

File: wave_optimization/scripts/plot_3d_data.py (hunk 3 above)
Comment (chutsu) [inline wave_optimization/scripts/plot_3d_data.py]: Scripts removed

File: docs/api/optimization/ceres/ceres_examples.md [hunk 7]
```diff
@@ -0,0 +1,73 @@
+# wave.ceres.ceres_examples
+
+This module contains code examples on how to use [Ceres-Solver][ceres-solver], in particular it contains exampls for how to implement a:
```
Comment (navganti) [inline docs/api/optimization/ceres/ceres_examples.md]: *examples. (Sorry for being pedantic, typos are my least favourite thing ever)

**Comment by stevenwaslander** [comment]
Can you add the PR checklist thingy and run through the steps? Also, probably need a review from someone who has used Ceres before.  My comments have been resolved.

File: docs/api/optimization/ceres/ceres_examples.md [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+
+    struct NumericalDiffCostFunctor {
+        template <typename T>
+        bool operator()(const T *const x, T *residual) const {
+            residual[0] = T(10.0) - x[0];
+            return true;
+        }
+    };
+
+    class AnalyticalCostFunction : public ::ceres::SizedCostFunction<1, 1> {
```
Comment (navganti) [inline docs/api/optimization/ceres/ceres_examples.md]: No "::" before ceres :).

File: docs/ref/optimizers/ceres_tutorial.md [hunk 9]
```diff
@@ -0,0 +1,205 @@
...
+- How to define a cost function?
+- How to define the optimization problem?
+- How do I define a numerical cost function?
+- How do I define an analytical cost function?
+- How do I load the optimization data?
+
+
+## How to define a cost function?
+Ceres an solve bounds constrained robustified non-linear least squares problems of the form:
```
Comment (navganti) [inline docs/ref/optimizers/ceres_tutorial.md]: can*

File: docs/ref/optimizers/ceres_tutorial.md [hunk 10]
```diff
@@ -0,0 +1,205 @@
...
+      template <typename T>
+      bool operator()(const T* const x, T* residual) const {
+        residual[0] = T(10.0) - x[0];
+        return true;
+      }
+    };
+
+Keypoints to the example code:
+
+- In Ceres a cost function has to be a functor (function objects: a plain old C++ object plus the `()` operator), this is essentially so that the ceres solver can call your cost function with the overridded `()` operator you implemented, this is where you will add your custom cost function, for more information on functors see [this][stack_overflow-functors].
```
Comment (navganti) [inline docs/ref/optimizers/ceres_tutorial.md]: run on sentence. After "implemented", add a period. :)

File: docs/ref/optimizers/ceres_tutorial.md [hunk 11]
```diff
@@ -0,0 +1,205 @@
...
+
+## How to define a cost function?
+Ceres an solve bounds constrained robustified non-linear least squares problems of the form:
+
+\begin{equation}
+  \min_{x} \dfrac{1}{2} \sum_{i} \rho_{i} (|| f_{i} (x_{i}, \dots, x_{i_{k}}) ||^{2})
+\end{equation}
+
+The expression $\rho_{i} (|| f_{i} (x_{i}, \dots, x_{i_{k}}) ||^{2})$ is known in ceres as a **`ResidualBlock`**, where $f_{i}(\cdot)$ is a cost function that depends on the parameter blocks $[x_{i}, \dots, x_{i_{k}}]$.
+
```
Comment (stevenwaslander) [inline docs/ref/optimizers/ceres_tutorial.md]: Do you have the cost function and residual backward?  A residual is usually the difference between expected and actual (so 10 and x), and the cost function is the norm squared and summed over all residuals.

File: docs/ref/optimizers/ceres_tutorial.md [hunk 12]
```diff
@@ -0,0 +1,205 @@
...
+\end{equation}
+
+This is a trivial problem, whose minimum is located at $x = 10$. The first step
+is to write a functor that will evaluate this function:
+
+    struct CostFunctor {
+      template <typename T>
+      bool operator()(const T* const x, T* residual) const {
+        residual[0] = T(10.0) - x[0];
+        return true;
```
Comment (stevenwaslander) [inline docs/ref/optimizers/ceres_tutorial.md]: The code looks right.

File: docs/ref/optimizers/ceres_tutorial.md [hunk 13]
```diff
@@ -0,0 +1,205 @@
...
+      }
+    };
+
+Keypoints to the example code:
+
+- In Ceres a cost function has to be a functor (function objects: a plain old C++ object plus the `()` operator), this is essentially so that the ceres solver can call your cost function with the overridded `()` operator you implemented, this is where you will add your custom cost function, for more information on functors see [this][stack_overflow-functors].
+- The output of your cost function is written to `residual[0]`, in all cases `residual` is always an array of type `T`, further in this example we are only writing to `residual[0]` because the output dimension for our scalar cost function is 1.
```
Comment (stevenwaslander) [inline docs/ref/optimizers/ceres_tutorial.md]: Try to be clear on the difference between cost function and residual throughout.  We only have to define the residual, because the cost function always is constructed the same way (cost = sum (residuals)^2).

File: docs/ref/optimizers/ceres_tutorial.md [hunk 14]
```diff
@@ -0,0 +1,205 @@
...
+                                  Dimension of residual -------------------+  |  |
+                                  Dimension of x -----------------------------+  |
+                                  Dimension of y --------------------------------+
+
+
+## How do I define an analytical cost function?
+The implementation for an analytical cost function is more involving compared to a numerical or automatic differentiated cost function. First we have to implement the analytical cost function using `ceres::SizedCostFunction`:
+
+    class QuadraticCostFunction
+      : public SizedCostFunction<1 /* number of residuals */,
```
Comment (stevenwaslander) [inline docs/ref/optimizers/ceres_tutorial.md]: Should this class be called AnalyticalCostFunction? Per the usage line below?

File: docs/ref/optimizers/ceres_tutorial.md [hunk 15]
```diff
@@ -0,0 +1,205 @@
...
+      }
+
+    private:
+      // Observations for a sample.
+      const double x;
+      const double y;
+    };
+
+The key-point to note is we are using the private `x` and `y` member variables to store an observation. Assuming the observations are in a `2n` sized array called `data` the problem construction is a simple matter of creating a `ceres::CostFunction` for every observation.
+
```
Comment (stevenwaslander) [inline docs/ref/optimizers/ceres_tutorial.md]: Is it just me, or does this seem wildly inefficient? A new heap function for every piece of data? Why not an array and a single residual function that runs over all data?

File: docs/ref/optimizers/ceres_tutorial.md [hunk 16]
```diff
@@ -0,0 +1,205 @@
...
+    ceres::Problem problem;
+    for (int i = 0; i < kNumObservations; ++i) {
+      ceres::CostFunction* cost_function =
+          new ceres::AutoDiffCostFunction<ExponentialResidual, 1, 1, 1>(
+              new ExponentialResidual(data[2 * i], data[2 * i + 1]));
+      problem.AddResidualBlock(cost_function, NULL, &m, &c);
+    }
+
+
```
Comment (stevenwaslander) [inline docs/ref/optimizers/ceres_tutorial.md]: Is it possible to include results, and maybe an easy way to plot them. so people know they did it right?

File: docs/ref/optimizers/ceres_tutorial.md [hunk 17]
```diff
@@ -0,0 +1,205 @@
...
+        // derivatives w.r.t. a subset of the parameter blocks.
+        if (jacobians != NULL && jacobians[0] != NULL) {
+          jacobians[0][0] = -1;
+        }
+        return true;
+      }
+    };
+
+Keypoints:
+
```
Comment (stevenwaslander) [inline docs/ref/optimizers/ceres_tutorial.md]: Love the keypoints, insanely informative.

File: scripts/install/ceres_install.bash [hunk 18]
```diff
@@ -1,15 +1,50 @@
...
+    git clone https://github.com/ceres-solver/ceres-solver.git
+
+    # go into ceres-solver repo and prepare for build
+    cd ceres-solver
+    mkdir build
+    cd build
+    cmake ..
+
+    # compile and install
+    make -j3
```
Comment (yuzhangbit) [inline scripts/install/ceres_install.bash]: Why `make -j3`?  Is there any reason of  not using `make -j$(nproc)` to automatically detect threads the machine has.

Comment (chutsu) [inline scripts/install/ceres_install.bash]: good point, but it shouldn't even be there. I'll leave it blank actually.

File: docs/ref/optimizers/ceres_tutorial.md (hunk 17 above)
Comment (chutsu) [inline docs/ref/optimizers/ceres_tutorial.md]: Thanks, those are things I found missing from the ceres website.

File: docs/ref/optimizers/ceres_tutorial.md (hunk 11 above)
Comment (chutsu) [inline docs/ref/optimizers/ceres_tutorial.md]: hmm, I actually copied it directly from the Ceres website: http://ceres-solver.org/nnls_tutorial.html#introduction. That is how they defined it, unless they are themselves wrong.

File: docs/api/optimization/ceres/ceres_examples.md (hunk 7 above)
Comment (chutsu) [inline docs/api/optimization/ceres/ceres_examples.md]: Done

File: docs/api/optimization/ceres/ceres_examples.md (hunk 8 above)
Comment (chutsu) [inline docs/api/optimization/ceres/ceres_examples.md]: Done

File: docs/ref/optimizers/ceres_tutorial.md (hunk 9 above)
Comment (chutsu) [inline docs/ref/optimizers/ceres_tutorial.md]: Done

File: docs/ref/optimizers/ceres_tutorial.md (hunk 13 above)
Comment (chutsu) [inline docs/ref/optimizers/ceres_tutorial.md]: Done

File: docs/ref/optimizers/ceres_tutorial.md (hunk 10 above)
Comment (chutsu) [inline docs/ref/optimizers/ceres_tutorial.md]: Done

File: docs/ref/optimizers/ceres_tutorial.md (hunk 14 above)
Comment (chutsu) [inline docs/ref/optimizers/ceres_tutorial.md]: Done

File: docs/ref/optimizers/ceres_tutorial.md (hunk 11 above)
Comment (stevenwaslander) [inline docs/ref/optimizers/ceres_tutorial.md]: Yeah, it became even more muddled below.  They refer to the residual function as the cost function, which it is, once you take the norm and square it.  Since that is the assumed form for all Ceres costs,  there is no cost function definition. They also do mention robust costs, which would imply they can change to a different cost, not just pure least squares. If you can make the distinction between residual and cost more clear in some places, that would be great.

File: docs/ref/optimizers/ceres_tutorial.md (hunk 15 above)
Comment (chutsu) [inline docs/ref/optimizers/ceres_tutorial.md]: From what I can tell they only have examples doing it that way, even a moderately complex example such as [solving the bundle adjustment problem ](https://ceres-solver.googlesource.com/ceres-solver/+/master/examples/bundle_adjuster.cc#260)uses this approach to load the data.

Comment (chutsu) [inline docs/ref/optimizers/ceres_tutorial.md]: I'll try and dig deeper and find out if this is true or not.

File: scripts/install/ceres_install.bash (hunk 18 above)
Comment (chutsu) [inline scripts/install/ceres_install.bash]: Done

File: wave_optimization/CMakeLists.txt (hunk 2 above)
Comment (chutsu) [inline wave_optimization/CMakeLists.txt]: Actually second thought, can we delay this for another PR? It's proving to be quite complicated

File: docs/ref/optimizers/ceres_tutorial.md (hunk 11 above)
Comment (chutsu) [inline docs/ref/optimizers/ceres_tutorial.md]: Having a closer look, I think they are just referring to the class `CostFunction` but not actually a real "cost function" in a mathematical sense. That is the only rational explanation I can come up with.

File: docs/ref/optimizers/ceres_tutorial.md (hunk 15 above)
Comment (stevenwaslander) [inline docs/ref/optimizers/ceres_tutorial.md]: No worries, something to keep in mind if we do run into efficiency issues.

**Comment by stevenwaslander** [comment]
Then the class name is wrong, pretty embarrassing. Clarify if possible.

File: docs/ref/optimizers/ceres_tutorial.md (hunk 11 above)
Comment (chutsu) [inline docs/ref/optimizers/ceres_tutorial.md]: Done

File: docs/ref/optimizers/ceres_tutorial.md (hunk 16 above)
Comment (chutsu) [inline docs/ref/optimizers/ceres_tutorial.md]: Included results, but not easy way to plot.

Comment (chutsu) [inline docs/ref/optimizers/ceres_tutorial.md]: Added plot script. Done

**Comment by chutsu** [comment]
I think I have addressed all issues in comments, please approve if you think its ready.

File: docs/ref/optimizers/ceres_tutorial.md [hunk 19]
```diff
@@ -90,7 +99,7 @@ Note that:
     Dimension of residual -----------------------------+  |
     Dimension of x ---------------------------------------+
 
-**Important**: In Ceres a cost function is not really a cost function in the mathematical sense, what you as the user is implementing is infact the residual because the cost function is always contructed the same way ($\text{cost} = \sum \text{residuals}^{2}$) (see "How do I define a cost function?" above).
+**Important**: In Ceres while implenting the `CostFunction` you as the user are expected to implement the **residual only** because the cost function is always contructed the same way ($\text{cost} = \sum \text{residuals}^{2}$) (see "How do I define a cost function?" above).
```
Comment (stevenwaslander) [inline docs/ref/optimizers/ceres_tutorial.md]: Looks like repeated text?

File: docs/ref/optimizers/ceres_tutorial.md [hunk 20]
```diff
@@ -117,7 +126,7 @@ Defining a numerical cost function is just a matter of using `ceres::NumericDiff
                                   Dimension of x -----------------------------+  |
                                   Dimension of y --------------------------------+
 
-**Important**: In Ceres a cost function is not really a cost function in the mathematical sense, what you as the user is implementing is infact the residual because the cost function is always contructed the same way ($\text{cost} = \sum \text{residuals}^{2}$) (see "How do I define a cost function?" above).
+**Important**: In Ceres while implenting the `CostFunction` you as the user are expected to implement the **residual only** because the cost function is always contructed the same way ($\text{cost} = \sum \text{residuals}^{2}$) (see "How do I define a cost function?" above).
```
Comment (stevenwaslander) [inline docs/ref/optimizers/ceres_tutorial.md]: Three times?

File: docs/ref/optimizers/ceres_tutorial.md [hunk 21]
```diff
@@ -9,22 +9,31 @@ As with all nonlinear-optimizers they can be difficult to master, in the followi
...
-Where $m$ is a set of data points, and function $f(x, \vec{\beta})$ that takes in input variable $x$ and $n$ parameters $\vec{\beta} = (\beta_1, \beta_2, \dots, \beta_n)$ with $m \geq n$. In the nonlinear least squares problem we seek to find the vector $\vec{\beta}$ that minimizes the residual $r$ between function $f(\cdot)$ output and desired output $d$ for every data point $i$.
-
 **Example** consider the problem of finding the minimum of the function:
 
 \begin{equation}
   \dfrac{1}{2} (10 - x)^{2}
 \end{equation}
 
-**Important**: In Ceres a cost function is not really a cost function in the mathematical sense, what you as the user is implementing is infact the residual because the cost function is always contructed the same way ($\text{cost} = \sum \text{residuals}^{2}$), in this case it is:
+**Important**: In Ceres while implenting the `CostFunction` you as the user are expected to implement the **residual only** because the cost function is always contructed the same way ($\text{cost} = \sum \text{residuals}^{2}$), in this case it is:
```
Comment (stevenwaslander) [inline docs/ref/optimizers/ceres_tutorial.md]: implenting (spelled wrong in all four copies.

**Comment by stevenwaslander** [comment]
Don't wreck my sidebar!
